mscore/preferences.h

```diff
@@ -66,156 +73,221 @@ enum class MusicxmlExportBreaks : char {
-      void init();
-      bool readDefaultStyle();
-      bool isThemeDark() { return globalStyle == MuseScoreStyleType::DARK_FUSION;}
+      ~Preferences();
+      void init(bool storeInMemoryOnly = false);
+      void save();
+
+      // general getters
+      QVariant defaultValue(const QString key) const;
+      bool getBool(const QString key) const;
+      QColor getColor(const QString key) const;
+      QString getString(const QString key) const;
+      int getInt(const QString key) const;
+      double getDouble(const QString key) const;
+
+      // general setters
+      void revertToDefaultValue(const QString key);
+      // set to true to let getters return default values instead of values from QSettings
+      void setReturnDefaultValues(bool returnDefaultValues);
+      void setPreference(const QString key, QVariant value);
+      // A temporary preference is stored "in memory" only and not written to file.
+      // If there is both a "normal" preference and a temporary preference with the same
+      // key the temporary preference is used
+      void setTemporaryPreference(const QString key, QVariant value);
+
+      /*
+       * Some preferences like enums and structs/classes are not easily read using the general set/get methods
+       * and therefore require specific getters and/or setters
+       */
+      SessionStart sessionStart() const;
+      MusicxmlExportBreaks musicxmlExportBreaks() const;
+      MuseScoreStyleType globalStyle() const;
+      bool isThemeDark() const;
+
+      template<typename T>
+      void setCustomPreference(const QString key, T t)
+            {
+            set(key, QVariant::fromValue<T>(t));
+            }
+
+      // The midiRemote preference requires special handling due to its complexity
+      MidiRemote midiRemote(int recordId) const;
+      void updateMidiRemote(int recordId, MidiRemoteType type, int data);
+      void clearMidiRemote(int recordId);
       };
 
-//---------------------------------------------------------
-//   ShortcutItem
-//---------------------------------------------------------
-
-class ShortcutItem : public QTreeWidgetItem {
+extern Preferences preferences;
 
-      bool operator<(const QTreeWidgetItem&) const;
+// Stream operators for enum classes
+// enum classes don't play well with QSettings without custom serialization
+template<typename T, typename std::enable_if<std::is_enum<T>::value>::type* = nullptr>
+inline QDataStream &operator<<(QDataStream &out, const T &val)
+{
+    return out << static_cast<int>(val);
+}
 
-   public:
-      ShortcutItem() : QTreeWidgetItem() {}
-      };
+template<typename T, typename std::enable_if<std::is_enum<T>::value>::type* = nullptr>
+inline QDataStream &operator>>(QDataStream &in, T &val)
+{
+    int tmp;
+    in >> tmp;
+    val = static_cast<T>(tmp);
+    return in;
+}
 
-extern Preferences preferences;
-extern bool useALSA, useJACK, usePortaudio, usePulseAudio;
 
 } // namespace Ms
+
+Q_DECLARE_METATYPE(Ms::SessionStart);
```

Comment:
Did you try with Q_ENUM instead ? https://doc.qt.io/qt-5/qobject.html#Q_ENUM 
My understanding is that we would not need to use template at all in this case.

Reply:
Yes, I tried using Q_ENUM but ran into some issues that led me to using templates.

Firstly, to use Q_ENUM the enum (or enum class) needs to declared within a QObject (or QGadget). The enums used for preferences are currently declared globally. Could they be moved to a class? If so which class?

The other problem i encountered was using enum class with QSettings gether.(https://stackoverflow.com/questions/22840660/enum-class-in-qvariant-in-qsettings). I didn't get it to work.

I will however make another attempt and share my issues. Maybe we can work it out together. I would love to remove those horrible templates.

Reply:
I tried again now using Q_ENUM_NS (for namespaces) which led to the same results as earlier attempts. See last commit on this branch: https://github.com/Soerboe/MuseScore/tree/try.using.qenum
You can check it out and see for yourself.

I get the following messages:
`
unknown:unknown: QVariant::load: unknown user type with name Ms::MuseScoreStyleType.
unknown:unknown: QVariant::load: unable to load type 1029.
/home/daghenning/MuseScore-development/MuseScore/mscore/musescore.cpp:main: set paper size from default printer
unknown:unknown: QVariant::load: unable to load type 1030.
unknown:unknown: QVariant::save: unable to save type 'Ms::SessionStart' (type id: 1029).

unknown:unknown: QVariant::save: unable to save type 'Ms::MusicxmlExportBreaks' (type id: 1030).

unknown:unknown: QVariant::save: unable to save type 'Ms::SessionStart' (type id: 1029).

unknown:unknown: QVariant::save: unable to save type 'Ms::MusicxmlExportBreaks' (type id: 1030).

unknown:unknown: QVariant::save: unable to save type 'Ms::SessionStart' (type id: 1029).

unknown:unknown: QVariant::save: unable to save type 'Ms::MusicxmlExportBreaks' (type id: 1030).

unknown:unknown: QVariant::save: unable to save type 'Ms::SessionStart' (type id: 1029).

unknown:unknown: QVariant::save: unable to save type 'Ms::MusicxmlExportBreaks' (type id: 1030).

unknown:unknown: QVariant::save: unable to save type 'Ms::SessionStart' (type id: 1029).

unknown:unknown: QVariant::save: unable to save type 'Ms::MusicxmlExportBreaks' (type id: 1030).
`

I have no idea how to cope with this... :(

Reply:
Excerpt from #qt IRC channel:

‎[18:19] ‎<‎Soerboe‎>‎ Does anyone have an example of how to store enum classes in QSettings? I can not get it to work. I try to use Q_ENUM_NS macro to register the enum class and use QVariant::fromValue<EnumType>(enumvalue) when using QSettings::setValue(). What I get is only the message "QVariant::save: unable to save type 'MyNameSpace::EnumType' (type id: 1030)".
‎[18:21] ‎<‎ericLemanissier‎>‎ @ddubya after studying sources, it seems bound values are cleared only when calling prepare, or exec(qstring)
‎[18:22] ‎<‎annulen‎>‎ Soerboe: did you register it with Q_ENUM or Q_ENUMS macro>
‎[18:22] ‎<‎annulen‎>‎ ?
‎[18:23] ‎<‎Soerboe‎>‎ annulen: Q_ENUM_NS since the enum class is declared in namespace outside of class
‎[18:24] ‎<‎ericLemanissier‎>‎ @Soerboe you should convert it to an int
‎[18:25] ‎<‎ericLemanissier‎>‎ QSettings can serialize/deserialize only a subset of QVariant types, and no user defined type
‎[18:25] ‎<‎ddubya‎>‎ ericLemanissier, that's why hes registering a new type
‎[18:26] ‎<‎ericLemanissier‎>‎ it does not give QSettings the ability to serialize it
‎[18:26] ‎<‎ericLemanissier‎>‎ https://code.woboq.org/qt5/qtbase/src/corelib/io/qsettings_win.cpp.html#674
‎[18:26] ‎<‎Soerboe‎>‎ ‎ericLemanissier‎: I had the impression that wasn't necessary when using Q_ENUM. That QVariant would handle that for me
‎[18:27] ‎<‎ddubya‎>‎ ericLemanissier, sure it does, QSettings uses QVariant
‎[18:27] ‎<‎ddubya‎>‎ the registered type is to get QVariant to regcognize it, not QSettings
‎[18:28] ‎<‎ddubya‎>‎ Soerboe, enums are always int (most compilers). The width might be less. so cast to (int) and your gold
‎[18:30] ‎<‎annulen‎>‎ Soerboe: or store string values obtained from QMetaEnum
‎[18:30] ‎<‎ericLemanissier‎>‎ maybe I'm wrong. It probably works if you qRegisterMetaTypeStreamOperators for your enum type
‎[18:32] ‎<‎Soerboe‎>‎ I tried using qRegisterMetaTypeStreamOperators but then I had to supply stream operators
‎[18:33] ‎<‎ericLemanissier‎>‎ sure,this is what is used to transform your type into bytes which can then be saved in QSettings
‎[18:34] ‎<‎Soerboe‎>‎ I works  by casting to int in the stream operators. I was just hoping there was a idiomatic Qt way of doing it.
‎[18:34] ‎<‎Soerboe‎>‎ :)
‎[18:38] ‎<‎ericLemanissier‎>‎ You have to decide how to perform the serialization. Qt cannot give a generic implementation, because it would cause problem if you modify the enum type after having stored it in qsettings, you would not necessary be able to read from the settings properly
‎[18:38] ‎<‎Soerboe‎>‎ So by using Q_ENUM the enum class can be stored in a QVariant, but to be able to store it in QSettings I have to supply a custom way of storing the enum, like casting to int or using string values from QMetaEnum. Am I right?
‎[18:39] ‎<‎ericLemanissier‎>‎ QSettings doc:Custom types registered using qRegisterMetaType() and qRegisterMetaTypeStreamOperators() can be stored using QSettings.
‎[18:42] ‎<‎ericLemanissier‎>‎ yes, you can either do what the doc say, or call settings::setValue(int(enumValue))

Reply:
Even if we keep the template, "store string values obtained from QMetaEnum" sounds promising no ? We could use this in the stream operator.

Thinking a bit differently. There is value in using enums vs simple strings (In a tabular view, the enum type could be combo) but Firefox about:config (which is more or less the goal of this change) only provide bool, int and string. Maybe we could get rid of the enum, use string instead and provide convenient getters.

Reply:
Yes, I will try to use QMetaEnum to save string representations of the enum values.

About your different thinking: If things were much easier using only boolean, int and string like in about:config I think that would be the way to go. But I do not think it is so much easier. Replacing an enum with a string requires "magic strings" to represent the legal values. Should these be just defines or strings stored in a data structure? What if an enum has 30 values. Then there would be a lot of overhead just to make sure a preference has a legal value. The string preference would be converted to a more convenient data structure eventually.

Reply:
One problem. The Q_NAMESPACE macro is needed for Q_ENUM_NS to work. But if it is used inside the same namespace in two different headers the linker complains about multiple definition of Ms::staticMetaObject. So it can't. Q_NAMESPACE is used in types.h in libmscore and therefore can not be used in preferences.h :(

How to fix? I can use a new namespace for preferences.h. Or use a new nested namespace inside the Ms namespace in preferences.h. What do you think?

(Maybe this should also be done in types.h?)

Reply:
I'm out of my confort zone now. @wschweer any opinion about this?

In theory, I would expect libmscore to have its own namespace and MuseScore another one...